boards/shared/Src/App/App_SharedSignal.c

```diff
@@ -1,135 +1,67 @@
-    signal->world                   = world;
-    signal->callback                = callback;
+    signal->is_callback_triggered = false;
 
-    return signal;
-}
-
-void App_SharedSignal_Destroy(struct Signal *signal)
-{
-    free(signal);
-}
+    App_Timer_InitTimer(signal->entry_timer, entry_time);
+    App_Timer_InitTimer(signal->exit_timer, exit_time);
 
-uint32_t App_SharedSignal_GetEntryLastTimeLowMs(const struct Signal *signal)
-{
-    return signal->entry_last_time_low_ms;
-}
-
-uint32_t App_SharedSignal_GetEntryLastTimeHighMs(const struct Signal *signal)
-{
-    return signal->entry_last_time_high_ms;
-}
-
-uint32_t App_SharedSignal_GetExitLastTimeLowMs(const struct Signal *signal)
-{
-    return signal->exit_last_time_low_ms;
-}
-
-uint32_t App_SharedSignal_GetExitLastTimeHighMs(const struct Signal *signal)
-{
-    return signal->exit_last_time_high_ms;
-}
-
-bool App_SharedSignal_IsCallbackTriggered(const struct Signal *const signal)
-{
-    return signal->is_callback_triggered;
+    return signal;
 }
 
-void App_SharedSignal_Update(struct Signal *signal, uint32_t current_time_ms)
+SignalState App_SharedSignal_Update(struct Signal *signal, bool entry_condition_high, bool exit_condition_high)
 {
-    if (signal->is_entry_condition_high(signal->world))
-    {
-        signal->entry_last_time_high_ms = current_time_ms;
-    }
-    else
-    {
-        signal->entry_last_time_low_ms = current_time_ms;
-    }
-
-    if (signal->is_exit_condition_high(signal->world))
-    {
-        signal->exit_last_time_high_ms = current_time_ms;
-    }
-    else
-    {
-        signal->exit_last_time_low_ms = current_time_ms;
-    }
+    TimerState entry_timer_state = App_Timer_RunIfCondition(signal->entry_timer, entry_condition_high);
+    TimerState exit_timer_state  = App_Timer_RunIfCondition(signal->exit_timer, exit_condition_high);
 
     if (!signal->is_callback_triggered)
     {
-        const uint32_t time_entry_condition_high_ms = current_time_ms - signal->entry_last_time_low_ms;
-
-        if (time_entry_condition_high_ms >= signal->callback.entry_condition_high_duration_ms)
+        if (entry_timer_state == TIMER_STATE_EXPIRED)
         {
-            signal->callback.function(signal->world);
+            signal->state                 = SIGNAL_ENTRY_HIGH;
             signal->is_callback_triggered = true;
         }
     }
     else
     {
-        const uint32_t time_exit_condition_high_ms = current_time_ms - signal->exit_last_time_low_ms;
-
-        if (time_exit_condition_high_ms >= signal->callback.exit_condition_high_duration_ms)
-        {
+        // keep running callback function for (exit_condition_high_duration_ms) time.
+        if (exit_timer_state == TIMER_STATE_EXPIRED)
             signal->is_callback_triggered = false;
-        }
         else
-        {
-            signal->callback.function(signal->world);
-        }
+            signal->state = SIGNAL_EXIT_HIGH;
     }
+    return signal->state;
```

Comment:
Some things to think about:
1. Don't need to pass in 2 arguments here: I believe `exit_condition_high` will always be the complement of `entry_condition_high` (if not you'll get some weird behavior)
2. What's the point of `signal->state`? Looks like it'll be set back to `SIGNAL_EXIT_HIGH` immediately after callback starts being triggered.
3. Don't think you need a `signal->state` variable and an `is_callback_triggered` variable -> I think they're supposed to give you the same info. You can remove `is_callback_triggered`. 
4. Maybe rename `SIGNAL_ENTRY/EXIT_HIGH` -> The names makes it sounds like they'll tell you whenever the entry or exit condition is high (which isn't true because there's the timer). Maybe `SIGNAL_STATE_ACTIVE/CLEARED`?

Reply:
You'll need to update the shared siganl tests too

Reply:
ok 2 is fixed
For 3, is_callback_triggered is the state function that we were discussing

Reply:
For 1, there are some things, such as brake and accelerator plausibility, which is activated with 25% accelerator and brake, and deactivated with <5% accelerator.

Reply:
> You'll need to update the shared siganl tests too

I will do the tests once the code is all reviewed

Reply:
For 1 you might get into weird behavior if enter and exit can ever be true simultaneously. maybe guard against this by only ticking entry timer if signal state inactive, and only ticking exit timer if signal state active.

Reply:
Also we're not doing callbacks anymore, so should probably rename

Reply:
> For 1 you might get into weird behavior if enter and exit can ever be true simultaneously. maybe guard against this by only ticking entry timer if signal state inactive, and only ticking exit timer if signal state active.

> Also we're not doing callbacks anymore, so should probably rename

> 4

Fixed with bf889d9c8cd8a129c8f54c7c1bcc674ce2fb78c3